packages/kit/src/runtime/app/forms.js

```diff
@@ -16,46 +16,57 @@ const ssr = import.meta.env.SSR;
 export const applyAction = ssr ? guard('applyAction') : client.apply_action;
 
 /** @type {import('$app/forms').enhance} */
-export function enhance(element, submit = () => {}) {
+export function enhance(form, submit = () => {}) {
 	/**
 	 * @param {{
-	 *   element: HTMLFormElement | HTMLButtonElement | HTMLInputElement;
-	 *   form: HTMLFormElement;
+	 *   action: string;
 	 *   result: import('types').ActionResult;
 	 * }} opts
 	 */
-	const fallback_callback = async ({ element, form, result }) => {
+	const fallback_callback = async ({ action, result }) => {
 		if (result.type === 'success') {
 			await invalidateAll();
 		}
 
-		const action = element.formAction ?? form.action;
-
 		if (location.origin + location.pathname === action.split('?')[0]) {
 			applyAction(result);
 		}
 	};
 
-	const form =
-		element instanceof HTMLFormElement ? element : /** @type {HTMLFormElement} */ (element.form);
-	if (!form) throw new Error('Element is not associated with a form');
-
 	/** @param {SubmitEvent} event */
 	async function handle_submit(event) {
 		event.preventDefault();
 
-		const action = element.formAction ?? form.action;
-
+		let action = form.action;
+		const element_action = /** @type {HTMLButtonElement | HTMLInputElement | null} */ (
+			event.submitter
+		)?.formAction;
+		// the browser will always set formAction - if not set, it defaults to the url;
+		// we therefore have to check that formAction is indeed set
+		if (action !== element_action && element_action) {
+			if (
+				// different urls
+				action.split('?')[0] !== element_action.split('?')[0] ||
+				// same url - formAction needs to be set to something or else we ignore it.
+				// we take advantage of the fact that there cannot be a default and named action
+				// at the same time.
+				element_action.includes('?')
+			) {
+				action = element_action;
+			}
+		}
 		const data = new FormData(form);
+		const controller = new AbortController();
 
 		let cancelled = false;
 		const cancel = () => (cancelled = true);
 
 		const callback =
 			submit({
-				element,
-				data,
+				action,
```

Comment:
Inside the fallback callback we are using the action name to know whether or not we should call `applyAction` . If people would want to do something similar they AFAIK have no way to do this without it, since the submit event with the potential formAction on it is gone by then. If they _have_ a way then we should omit this